File: tkpdinbox/src/main/java/com/tokopedia/inbox/inboxchat/chattemplate/view/adapter/TemplateChatTypeFactoryImpl.java
package com.tokopedia.inbox.inboxchat.chattemplate.view.adapter;

import android.view.View;

import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory;
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder;
import com.tokopedia.inbox.inboxchat.chatroom.fragment.ChatRoomFragment;
import com.tokopedia.inbox.inboxchat.chatroom.listener.ChatRoomContract;
import com.tokopedia.inbox.inboxchat.chattemplate.view.adapter.viewholder.TemplateChatViewHolder;
import com.tokopedia.inbox.inboxchat.chattemplate.view.viewmodel.TemplateChatModel;

/**
 * Created by stevenfredian on 9/27/17.
 */

public class TemplateChatTypeFactoryImpl extends BaseAdapterTypeFactory implements TemplateChatTypeFactory{

    ChatRoomContract.View viewListener;

    public TemplateChatTypeFactoryImpl(ChatRoomFragment context) {
        this.viewListener = context;
    }

    @Override
    public AbstractViewHolder createViewHolder(View view, int type) {

        AbstractViewHolder viewHolder;

        if (type == TemplateChatViewHolder.LAYOUT)
            viewHolder = new TemplateChatViewHolder(view, viewListener);
        else
            return super.createViewHolder(view, type);

        return viewHolder;
    }

    @Override
    public int type(TemplateChatModel templateChatModel) {
        return TemplateChatViewHolder.LAYOUT;
    }

}

File: features/communication/topchat/src/main/java/com/tokopedia/topchat/chattemplate/view/adapter/TemplateChatTypeFactoryImpl.kt
package com.tokopedia.topchat.chattemplate.view.adapter

import android.view.View
import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder
import com.tokopedia.topchat.chattemplate.view.adapter.viewholder.TemplateChatViewHolder
import com.tokopedia.topchat.chattemplate.view.listener.ChatTemplateListener
import com.tokopedia.topchat.chattemplate.view.uimodel.TemplateChatUiModel

class TemplateChatTypeFactoryImpl(
    var viewListener: ChatTemplateListener
) : BaseAdapterTypeFactory(), TemplateChatTypeFactory {

    override fun createViewHolder(view: View, viewType: Int): AbstractViewHolder<*> {
        return if (viewType == TemplateChatViewHolder.LAYOUT)
            TemplateChatViewHolder(view, viewListener)
        else return super.createViewHolder(view, viewType)
    }

    override fun type(templateChatUiModel: TemplateChatUiModel): Int {
        return TemplateChatViewHolder.LAYOUT
    }
}